File: C++/DesignPatterns/SimpleFactoryDesignPattern.cpp
#include <iostream>
#include <memory>
#include <stdexcept>
#include <string>

struct Notification {
    virtual void send(const std::string& message) = 0;
    virtual ~Notification() = default;
};

struct SMSNotification : Notification {
    void send(const std::string& message) override {
        throw std::runtime_error("Unimplemented method 'send'");
    }
};

struct SlackNotification : Notification {
    void send(const std::string& message) override {
        throw std::runtime_error("Unimplemented method 'send'");
    }
};

struct WhatsappNotification : Notification {
    void send(const std::string& message) override {
        throw std::runtime_error("Unimplemented method 'send'");
    }
};

class NotificationFactory {
public:
    static std::unique_ptr<Notification> getNotification(const std::string& type) {
        if (type == "SMS") {
            return std::make_unique<SMSNotification>();
        } else if (type == "SLACK") {
            return std::make_unique<SlackNotification>();
        } else if (type == "Whatsapp") {
            return std::make_unique<WhatsappNotification>();
        }
        throw std::invalid_argument("Unknown notification type: " + type);
    }
};

class Application {
public:
    void sendNotification() {
        auto notification = NotificationFactory::getNotification("SMS");
        notification->send("message");
    }
};

// Simple Factory Design Pattern

int main() {
    Application application;
    application.sendNotification();
    return 0;
}

File: Python/DesignPatterns/SimpleFactoryDesignPattern.py
from abc import ABC, abstractmethod


class Notification(ABC):
    @abstractmethod
    def send(self, message: str) -> None:
        raise NotImplementedError


class SMSNotification(Notification):
    def send(self, message: str) -> None:
        raise NotImplementedError("Unimplemented method 'send'")


class SlackNotification(Notification):
    def send(self, message: str) -> None:
        raise NotImplementedError("Unimplemented method 'send'")


class WhatsappNotification(Notification):
    def send(self, message: str) -> None:
        raise NotImplementedError("Unimplemented method 'send'")


class NotificationFactory:
    @staticmethod
    def get_notification(notification_type: str) -> Notification:
        if notification_type == "SMS":
            return SMSNotification()
        elif notification_type == "SLACK":
            return SlackNotification()
        elif notification_type == "Whatsapp":
            return WhatsappNotification()
        else:
            raise ValueError(f"Unknown notification type: {notification_type}")


class Application:
    def send_notification(self) -> None:
        notification = NotificationFactory.get_notification("SMS")
        notification.send("message")


# Simple Factory Design Pattern


def main() -> None:
    Application().send_notification()


if __name__ == "__main__":
    main()
